### PR TITLE
[WIP]Add Progress bar for artifacts download & upload

### PR DESCRIPTION
**Pull request by serena-ruan**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?
Add progress bar for artifacts upload & download

Option 1: Using tqdm.notebook.tqdm
This definitely has some displaying issue as it doesn't make to 100% even when the process is completed. The completed progress bar doesn't disappear as well.

https://github.com/mlflow/mlflow/assets/82044803/b914b3f3-d37b-425b-87bc-ad7512cd8fa3

Option 2: Using tqdm.tqdm
This displays normally, it's just not that fancy :) transformers package uses this way as well.

https://github.com/mlflow/mlflow/assets/82044803/ada7b1f2-78e2-4706-8db6-355904db1c91

But still, it has a problem on Databricks notebook. In order to keep the in-line updates of a progress bar (instead of displaying lots of progress bars showing the progress increasing), we need to add config `position=0`. But seems it's displaying first and then truncate the result, so when the output reaches Databricks's notebook output limit, it truncates at a strange place:

https://github.com/mlflow/mlflow/assets/82044803/b5d3f0c7-5c85-4580-b98c-bdb430a61581

## How is this patch tested?

Tested in Databricks.

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by WeichenXu123** [comment]
Could you record a demo video and attach it to the PR description ? :)

File: mlflow/store/artifact/artifact_repo.py [hunk 1]
```diff
@@ -189,7 +190,8 @@ def _download_file(src_artifact_path, dst_local_dir_path):
 
         # Wait for downloads to complete and collect failures
         failed_downloads = {}
-        for f in as_completed(futures):
+        for f in (pbar := tqdm(as_completed(futures), total=len(futures))):
+            pbar.set_description(f"Downloading artifact {futures[f]}")
```
Comment (harupy) [inline mlflow/store/artifact/artifact_repo.py:194]:
```suggestion
        pbar = tqdm(as_completed(futures), total=len(futures))
        for f in pbar:
            pbar.set_description(f"Downloading artifact {futures[f]}")
```

I'd use a simple assignment here.

File: mlflow/store/artifact/databricks_artifact_repo.py [hunk 2]
```diff
@@ -88,12 +89,30 @@ def _complete_futures(futures_dict):
...
+                results[key] = future.result()
+                pbar.update(
+                    min(
+                        file_size - index * _MULTIPART_UPLOAD_CHUNK_SIZE,
+                        _MULTIPART_UPLOAD_CHUNK_SIZE,
+                    )
+                )
+                pbar.refresh()
+            except Exception as e:
+                errors[key] = repr(e)
```
Comment (harupy) [inline mlflow/store/artifact/databricks_artifact_repo.py:115]: I filed https://github.com/mlflow/mlflow/pull/8570. Once it's merged, we can pull out this logic from `_complete_futures`.

**Comment by harupy** [comment]
@serena-ruan Can you take a screen recording and attach it in the PR description?

**Comment by serena-ruan** [comment]
> @serena-ruan Can you take a screen recording and attach it in the PR description?

Yes, I'm still testing, the behavior of tqdm is quite strange, sometimes it doesn't make it to 100% and even when it does, and I added the config `leave=False` which is supposed to remove the progress bar after complete, but it doesn't work well. Once I make sure it works I'll update the PR description.

**Comment by harupy** [comment]
Does the same thing happen on Jupyter notebook?

**Comment by serena-ruan** [comment]
I'll reopen it after testing :) Don't want to trigger the workflow tests

**Comment by harupy** [comment]
You can add `[skip ci]` to a commit message to skip workflow runs:

https://docs.github.com/en/actions/managing-workflow-runs/skipping-workflow-runs

**Pull request closed without merging**